### PR TITLE
[BEAM-6655] Initial support for progress reporting over the FnAPI.

### PR DESCRIPTION
**Pull request by robertwb**
This uses the existing FnAPI protos which may need revision as the design is finalized.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by robertwb** [comment]
The first commit is in pending https://github.com/apache/beam/pull/7801 (which would otherwise cause conflicts) so this PR is focused on the second commit. 

R: @chamikaramj @ananvay 
CC: @mxm

File: sdks/python/apache_beam/runners/worker/bundle_processor.py [hunk 1]
```diff
@@ -148,8 +148,13 @@ def try_split(self, fraction_of_remainder, total_buffer_size):
         # We are "finished" with the (non-existent) previous element.
         current_element_progress = 1
       else:
-        # TODO(SDF): Get actual progress of current element.
-        current_element_progress = 0.5
+        current_element_progress_object = (
+            self.receivers[0].current_element_progress())
+        if current_element_progress_object is None:
+          current_element_progress = 0.5
```
Comment (mxm) [inline sdks/python/apache_beam/runners/worker/bundle_processor.py]: Shouldn't we return `0.0` or a special value like `-1` to indicate progress is unknown?

Comment (robertwb) [reply sdks/python/apache_beam/runners/worker/bundle_processor.py]: We don't return this value, but use it in the computations below. We know we're *somewhere* in the element, so half way is a reasonable choice in the absence of more specific data.

File: sdks/python/apache_beam/io/iobase.py [hunk 2]
```diff
@@ -1165,3 +1171,56 @@ def check_done(self):
...
+
+
+class RestrictionProgress(object):
+  """Used to record the progress of a restriction.
+
+  Experimental; no backwards-compatibility guarantees.
+  """
+  def __init__(self, **kwargs):
+    # Only accept keyword arguments.
+    self._fraction = kwargs.pop('fraction', None)
```
Comment (ananvay) [inline sdks/python/apache_beam/io/iobase.py]: Should we have this be a cumulative count of number of elements seen so far, or a fraction?

File: sdks/python/apache_beam/io/iobase.py [hunk 3]
```diff
@@ -1165,3 +1171,56 @@ def check_done(self):
...
+    self._completed = kwargs.pop('completed', None)
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/io/iobase.py]: No, this has nothing to do with split points. Renamed to completed_work and remaining_work.

File: sdks/python/apache_beam/io/iobase.py [hunk 4]
```diff
@@ -1165,3 +1171,56 @@ def check_done(self):
...
+
+  @property
+  def remaining(self):
+    if self._remaining:
+      return self._remaining
+    elif self._completed:
+      return self._completed * (1 - self._fraction) / self._fraction
+
+  @property
+  def total(self):
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/io/iobase.py]: Done.

File: sdks/python/apache_beam/io/restriction_trackers.py [hunk 5]
```diff
@@ -102,6 +103,16 @@ def current_restriction(self):
...
     return self._current_watermark
 
+  def current_progress(self):
+    with self._lock:
+      if self._current_position is None:
+        fraction = 0.0
+      else:
+        fraction = (
+            float(self._current_position - self._range.start)
+            / (self._range.stop - self._range.start))
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/io/restriction_trackers.py]: Changed to return a 1 in this case.

File: sdks/python/apache_beam/runners/worker/operations.py [hunk 6]
```diff
@@ -106,6 +107,11 @@ def try_split(self, fraction_of_remainder):
     # of the consumers separately.
     return None
 
+  def current_element_progress(self):
+    # TODO(SDF): Could implement this as a weighted average, if it becomes
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/runners/worker/operations.py]: Added documentation.

File: sdks/python/apache_beam/runners/worker/operations.py [hunk 7]
```diff
@@ -584,19 +596,57 @@ def monitoring_infos(self, transform_id):
...
-      delayed_application = self.dofn_runner.process_with_restriction(o)
-      if delayed_application:
-        self.execution_context.delayed_applications.append(
-            (self, delayed_application))
+      try:
+        with self.lock:
+          self.element_start_output_bytes = self._total_output_bytes()
+          for receiver in self.tagged_receivers.values():
+            receiver.opcounter.restart_sampling()
+        delayed_application = self.dofn_runner.process_with_restriction(o)
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/runners/worker/operations.py]: Done.

**Comment by robertwb** [comment]
Run Python PreCommit

**Comment by ananvay** [comment]
lgtm, thanks Robert!

**Comment by chamikaramj** [comment]
LGTM. Thanks.

**Pull request merged**